### DISCUSSION TITLE
Null-safe dereference prevents out var from being definitely assigned even when result is non-null

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#1195] [General]
I would expect `a` to be definitely assigned since `== true` implies `!= null` which means `TryA` was definitely executed [(sharplab)](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiUBmLAJgwCEiBvIjJrc4AezYBsMAVBATwCCACjYBXGBgCWAO0kQAlI2YNCzdRggYAvBgAsFAmo1MUAdgwwEYuEfUBfIo+KEylDAGF6y0+Wo0MOgwAczgYXAxndR8WLD0MAFlhJWMmVRMmKQAzDGFqAH4AOj4hUQkMADcIBE0FHV0rGxSMwJiWnABOYUU7DKjmZ3sgA):

```cs
using System;

public class B
{
    public bool TryA(out int a)
    {
        a = 42;
        return true;
    }
}

public class C
{
    public B B { get; }
    
    public void M()
    {
        if (B?.TryA(out var a) == true)
        {
            Console.WriteLine(a);
        }
    }
}
```

Is there a good reason for this or is it just not implemented yet?

**Comment by yaakov-h** [comment]
I assume this is applicable? https://github.com/dotnet/roslyn/issues/12514#issuecomment-233206340

**Comment by jnm2** [comment]
@gafter

> Although you might like the compiler to do better here, the behavior is as specified for the `?.` operator and the `??` operator. The `??` operator does not specify any special behavior for _definitely-assigned-when-true_ and _definitely-assigned-when-false_ when its result happens to be of type `bool`.

This is not specific to bool though. Reference types are the same way. So long as the comparison you do implies `!= null`, the type should be irrelevant. [(example)](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiUBmLAJgwCEiBvIjJrcnTAFQQE8BBACgD2AVxgYAlgDtREAJSNmDQs2UYIGALwYALBQJKVTFAHYMAIh2m9ygL5FbxQmUoYAwvXmHy1GhjoYA5nAwuBj2yh4sWFoYALJ8cvpMigZMYgBmGHzUAPwAdJy8giIYAG4QCKoyGprmFKYJKb4RjTgAnHyyVilhzPbWQA)

**Comment by gafter** [comment]
@jnm2 Right, there is also no special behavior specified for *definitely assigned when non-null*.

**Comment by jnm2** [comment]
@gafter Cool. Is there any reason we couldn't have it?

**Comment by gafter** [comment]
@jnm2 Right now “it” is mythical. It would be very complex to specify, implement, and test, for little value to programmers.

**Comment by NickCraver** [comment]
I think we may want to revisit this with nullable changes proposed for C# 8, still hitting this fairly often.
